src/Microsoft.ApplicationInsights/Managed/Shared/Extensibility/Implementation/Tracing/HeartbeatProvider.cs

```diff
@@ -0,0 +1,346 @@
+﻿namespace Microsoft.ApplicationInsights.Extensibility.Implementation.Tracing
+{
+    using System;
+    using System.Collections.Concurrent;
+    using System.Collections.Generic;
+    using System.Globalization;
+    using System.Linq;
+    using System.Threading;
+    using Microsoft.ApplicationInsights.DataContracts;
+    using Microsoft.ApplicationInsights.Extensibility;
+
+    /// <summary>
+    /// Implementation of health heartbeat functionality.
+    /// </summary>
+    internal class HeartbeatProvider : IDisposable, IHeartbeatProvider
+    {
+        /// <summary>
+        /// The default interval between heartbeats if not specified by the user. Left public for use in unit tests.
+        /// </summary>
+        public static int DefaultHeartbeatIntervalMs = 5000;
+
+        protected List<string> disabledDefaultFields; // string containing fields that are not to be sent with the payload. null means send everything available.
+
+        /// <summary>
+        /// The name of the health heartbeat metric item and operation context. 
+        /// </summary>
+        private static string heartbeatSyntheticMetricName = "HealthState";
+
+        /// <summary>
+        /// The SDK supplied 'default' payload items to send out with each health heartbeat.
+        /// </summary>
+        private ConcurrentDictionary<string, HeartbeatPropertyPayload> sdkPayloadItems;
+
+        /// <summary>
+        /// The extended payload items to send out with each health heartbeat.
+        /// </summary>
+        private ConcurrentDictionary<string, HeartbeatPropertyPayload> extendPayloadItems;
+
+        private bool disposedValue = false; // To detect redundant calls to dispose
+        private TimeSpan heartbeatInterval; // time between heartbeats emitted specified in milliseconds
+        private TelemetryClient telemetryClient; // client to use in sending our heartbeat
+        private int heartbeatsSent; // counter of all heartbeats
+
+        public HeartbeatProvider() : this(TimeSpan.FromMilliseconds(DefaultHeartbeatIntervalMs), null)
+        {
+        }
+
+        public HeartbeatProvider(TimeSpan heartbeatInterval, IEnumerable<string> disabledDefaultFields)
+        {
+            this.disabledDefaultFields = disabledDefaultFields?.ToList();
+            this.heartbeatInterval = heartbeatInterval;
+            this.extendPayloadItems = new ConcurrentDictionary<string, HeartbeatPropertyPayload>(StringComparer.OrdinalIgnoreCase);
+            this.sdkPayloadItems = new ConcurrentDictionary<string, HeartbeatPropertyPayload>(StringComparer.OrdinalIgnoreCase);
+            this.heartbeatsSent = 0; // count up from construction time
+        }
+
+        /// <summary>
+        /// Gets the currently defined interval between heartbeats (primarily used in unit testing scenarios)
+        /// </summary>
+        public TimeSpan HeartbeatInterval => this.heartbeatInterval;
+
+        public string DiagnosticsInstrumentationKey { get; set; }
```

Comment:
Do we envision cases where we need to track heartbeat to the two different destinations? We were asked many times to provide a way to send telemetry to two destinations (e.g. in "provider" or "host" cases), so some SDK consumers already have code to send to two different Telemetry Configurations or two different clients with their own IKeys.
Also, Diagnostics Telemetry Module provides an override to forward telemetry to a separate diagnostics Ikey, it looks like HeartBeat will be forwarded there as well. If heartbeat is set for alerting, debugging on the deployment and forwarding data elsewhere will trigger the "no hearbeat" alert in the main AI resource.